Add network attachment definition creation form

pcbailey opened the pull request:
This PR adds a form for creating network attachment definitions.

![Selection_614](https://user-images.githubusercontent.com/8544299/67058388-9acfc300-f122-11e9-801a-24674d39e252.png)

Linux bridge screenshot:
![OKD - Google Chrome_617](https://user-images.githubusercontent.com/8544299/67492154-fbd82900-f643-11e9-8262-85a2d68bb329.png)

SR-IOV screenshot 1:
![OKD - Google Chrome_618](https://user-images.githubusercontent.com/8544299/67502394-d0f5d100-f653-11e9-9c55-1df9d96450e4.png)

SR-IOV screenshot 2 (showing the "Resource Name" dropdown opened):
![OKD - Google Chrome_619](https://user-images.githubusercontent.com/8544299/67502426-dd7a2980-f653-11e9-911e-57a1fc357698.png)

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/hooks/network-specific-elements-hook.tsx:
```diff
@@ -0,0 +1,174 @@
+import { Dropdown } from '@console/internal/components/utils';
+import { ELEMENT_TYPES, NetworkTypeParams, networkTypeParams } from '../constants';
+
+export const useNetworkSpecificElements = (
```
hook which returns JSX ? thats not correct, maybe you want to use `React.useCallback` directly in `network-attachment-definitions-form.tsx `

pcbailey replied:
`React.useCallback` worked like a charm. =)

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/hooks/form-validation-hook.ts:
```diff
@@ -0,0 +1,40 @@
+import * as _ from 'lodash';
+import { networkTypeParams } from '../constants';
+
+export const useFormValidationHook = (fieldErrors, name, networkType, typeParamsData) => {
```
this does not have to be custom hook, it can be just a function which will be used in `React.useCallback`

pcbailey replied:
As usual, you're right. =) Done!

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/network-attachment-definitions-form.tsx:
```diff
@@ -0,0 +1,252 @@
+export const NetworkAttachmentDefinitionForm_ = (props) => {
+  const { loaded, match, resources } = props;
+  const namespace = getNamespace(match);
+  const sriovNetNodePoliciesData = getSriovNetNodePolicyData(resources);
```
form is missing error handling in case `sriovnetworknodepolicies` cannot be loaded

pcbailey replied:
Done!

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/network-attachment-definitions-form.tsx:
```diff
@@ -0,0 +1,252 @@
+import { NetworkAttachmentDefinitionConfig } from '../../types';
+import { useFormValidationHook, useNetworkSpecificElements } from '../../hooks';
+
+export const networkTypes = {
```
Done!

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/network-attachment-definitions-form.tsx:
```diff
@@ -0,0 +1,252 @@
+  bridge: 'Linux bridge',
+};
+
+// eslint-disable-next-line no-underscore-dangle
```
Done! I also renamed the files for components and added default exports for each of them.

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/network-attachment-definitions-form.tsx:
```diff
@@ -0,0 +1,252 @@
+};
+
+// eslint-disable-next-line no-underscore-dangle
+export const NetworkAttachmentDefinitionForm_ = (props) => {
```
Done!

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/network-attachment-definitions-form.tsx:
```diff
@@ -0,0 +1,252 @@
+  );
+};
+
+export type NetworkAttachmentDefinitionFormProps = {
```
Done!

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/network-attachment-definitions-form.tsx:
```diff
@@ -0,0 +1,252 @@
+)(NetworkAttachmentDefinitionForm_);
+
+export const NetworkAttachmentDefinitionForm = (props) => {
+  const resources = [
```
Done!

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/network-attachment-definitions-form.tsx:
```diff
@@ -0,0 +1,252 @@
+
+export const NetworkAttachmentDefinitionForm = (props) => {
+  const resources = [
+    {
```
That's because it wasn't. =) I've removed it.

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/network-attachment-definitions-form.tsx:
```diff
@@ -0,0 +1,252 @@
+// eslint-disable-next-line no-underscore-dangle
+export const NetworkAttachmentDefinitionForm_ = (props) => {
+  const { loaded, match, resources } = props;
+  const namespace = getNamespace(match);
```
Done!

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,433 @@
+    return types;
+  }, [hasSriovNetNodePolicyCRD]);
+
+  const buildConfig = (): NetworkAttachmentDefinitionConfig => {
```
it would be better to move these functions outside of component to not recreate them on every refresh. Applies for the ones below too

pcbailey replied:
Done!

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,433 @@
+    return validateForm();
+  }, [fieldErrors, name, networkType, typeParamsData]);
+
+  const getNetworkTypes = React.useCallback(() => {
```
I dont think we gain anything by having this inside the component and as a hook. I'd move it outside and then you could memoize it (React.useMemo)

pcbailey replied:
Done!

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,433 @@
+    return getTypeElements();
+  }, [networkType, typeParamsData, setTypeParamsData, sriovNetNodePoliciesData]);
+
+  const formIsValid = React.useCallback(() => {
```
I'd move it outside the component and also having nested functions is unnecessary IMO, they are recreated everytime the function is called

pcbailey replied:
Done!

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,433 @@
+      );
+    };
+
+    return getTypeElements();
```
why returning the result of `getTypeElements()` when you can have a function body here directly' ?

pcbailey replied:
Fixed!

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,433 @@
+  const [fieldErrors, setFieldErrors] = React.useState<FieldErrors>({});
+
+  const getNetworkTypeSpecificElements = React.useCallback(() => {
+    const handleTypeParamChange = (paramKey, event, elemType) => {
```
all functions inside function are recreated everytime the parent function is called. Can you please move them outside ?

pcbailey replied:
Done!

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/constants/index.ts:
```diff
@@ -1 +1,66 @@
+  name: string;
+  required?: boolean;
+  type: string;
+  validation?: (...args: any[]) => string | null;
```
cant we have something better than `any` ?

pcbailey replied:
I just removed it. None of the current networks have field validation at the moment, so it's not necessary. I'll add better typing if I end up adding validation for parameters later on.

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,433 @@
+  },
+];
+
+export const NetworkAttachmentDefinitionForm = (props) => {
```
Only exporting it as the default causes issues with the call to `import()` in the loader in the plugin file.

rawagner replied:
for default export you can do `m.default` instead of `m.NetworkAttachmentDefinitionForm`

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,433 @@
+};
+NetworkAttachmentDefinitionForm.displayName = 'NetworkAttachmentDefinitionForm';
+
+export type NetworkAttachmentDefinitionFormProps = {
```
Done!

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,433 @@
+
+export const ConnectedNetworkAttachmentDefinitionForm = connect(
+  mapStateToProps,
+  mapDispatchToProps,
```
Done!

pcbailey commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,433 @@
+  const k8sModels = k8s.getIn(['RESOURCES', 'models']);
+
+  return {
+    k8s,
```
Done!

phoracek commented:
@pcbailey I think we can drop the "L2 if no IP mgmt specified label". The field does not have a star, hence it is not mandatory. If you don't specify anything, you get no IP management. But up to you. Nothing bad with the current state.

Could you maybe just share a screen shot with the IP management drop down opened?

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,488 @@
+
+  const config = JSON.stringify(buildConfig(name, networkType, typeParamsData));
+  const newNetAttachDef = {
+    apiVersion: 'k8s.cni.cncf.io/v1',
```
it would be better to get this from `NetworkAttachmentDefinitionModel` such as NetworkAttachmentDefinitionModel.apiGroup/NetworkAttachmentDefinitionModel.apiVersion

pcbailey replied:
Done!

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,488 @@
+  setLoading,
+) => {
+  e.preventDefault();
+
```
I think you should also call setError(null) to make sure that the previous error is not shown anymore

pcbailey replied:
Done!

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,488 @@
+
+  k8sCreate(NetworkAttachmentDefinitionModel, newNetAttachDef)
+    .then(() => {
+      setLoading(false);
```
i'd put this to `finally` block

pcbailey replied:
I added `setLoading(false)` to the `catch` block as discussed.

rawagner commented on frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx:
```diff
@@ -0,0 +1,488 @@
+  return resourceNames;
+};
+
+const getNetworkTypeSpecificElements = (
```
IMO this should be a component

pcbailey replied:
Done!

pcbailey commented:
@phoracek I was under the impression that the `IP Address Management` field is supposed to be a multi-line text field, not a dropdown. Did you mean a different field or did I misunderstand the design?

phoracek commented:
@pcbailey it should be, but in the screenshot on the cover message of this PR, there is this little :arrow_down_small: . Is it just to expand the text field?

pcbailey commented:
@phoracek I'm sorry. Somehow that field got set to dropdown at some point. I fixed it a couple of days ago and didn't even think to check the screenshots. I just updated the screenshot. =)

phoracek commented:
@pcbailey could you share a screenshot of SR-IOV too?

pcbailey commented:
@phoracek Done. =)

rawagner commented:
/lgtm

But create a followup to improve types, there are quite a few places without them
